HW0/P4/P4.py

```diff
@@ -31,46 +32,77 @@
     # Normally, this data wouldn't be available in the real world
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '--b', label='True trajectory')
+    s_true = np.transpose(np.loadtxt('P4_trajectory.txt', delimiter=','))
+    ax.plot(s_true[0,:], s_true[1,:], s_true[2,:],
+            '--b', label='True trajectory')
 
     #####################
     # Part 2:
     #
     # Read the observation array and plot it (Part 2)
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    m = np.transpose(np.loadtxt('P4_measurements.txt', delimiter=','))
+    ax.plot(m[0,:] / rx, m[1,:] / ry, m[2,:] / rz,
+            '--g', label='True trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    A = np.matrix([[1, 0, 0, dt, 0, 0], [0, 1, 0, 0, dt, 0], [0, 0, 1, 0, 0, dt], [0, 0, 0, 1 - c * dt, 0, 0], [0, 0, 0, 0, 1 - c * dt, 0], [0, 0, 0, 0, 0, 1 - c * dt]])
+    a = np.matrix(np.zeros((6, 1)))
+    a[5] = g * dt
+    s = np.matrix(np.zeros((6, K)))
+
+    s[:, 0] = np.transpose([[0, 0, 2, 15, 3.5, 4.0]])
+    for i in xrange(1, K, 1):
+        s[:, i] = A * s[:, i - 1] + a
 
     # Initial conditions for s0
     # Compute the rest of sk using Eq (1)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    s = np.array(s)
+    ax.plot(s[0,:], s[1,:], s[2,:],
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.matrix([[bx, 0, 0, 0, 0, 0], [0, by, 0, 0, 0, 0], [0, 0, bz, 0, 0, 0], [0, 0, 0, bvx, 0, 0], [0, 0, 0, 0, bvy, 0], [0, 0, 0, 0, 0, bvz]])
+    C = np.matrix([[rx, 0, 0, 0, 0, 0], [0, ry, 0, 0, 0, 0], [0, 0, rz, 0, 0, 0]])
 
     # Initial conditions for s0 and Sigma0
+    s = np.matrix(np.zeros((6, K)))
+    s[:, 0] = np.transpose([[0, 0, 2, 15, 3.5, 4.0]])
+    sigma = 0.01 * np.matrix(np.identity(6))
+    m = np.matrix(m)
+
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    def predictS(sk):
+        return A * sk + a
+
+    def predictSig(sigk):
+        return np.linalg.pinv(A * sigk * np.transpose(A) + B * np.transpose(B))
```

Comment:
It seems that we get different results in the predictSig function, but not highly important (might be the use of pinv instead of inv).